File: python/socketd/socketd/transport/core/impl/ConfigBase.py
import os
import ssl
from abc import ABC
from typing import Optional
from concurrent.futures import ThreadPoolExecutor

from socketd.exception.SocketDExecption import SocketDException
from socketd.transport.core.Asserts import Asserts
from socketd.transport.core.Config import Config
from socketd.transport.core.Codec import Codec
from socketd.transport.core.IdGenerator import IdGenerator, GuidGenerator
from socketd.transport.core.codec.CodecDefault import CodecDefault
from socketd.transport.core.Costants import Constants
from socketd.transport.core.fragment.FragmentHandlerDefault import FragmentHandlerDefault

from socketd.utils.LogConfig import logger
from socketd.transport.stream.StreamManger import StreamManger
from socketd.transport.stream.StreamMangerDefault import StreamMangerDefault


class ConfigBase(Config, ABC):

    def __init__(self, clientMode: bool):
        self._clientMode = clientMode
        self._serialSend = False
        self._nolockSend = False

        self._streamManger: StreamManger = StreamMangerDefault(self)
        self._codec: Codec = CodecDefault(self)

        self._charset = "utf-8"

        self._idGenerator: type[GuidGenerator] = GuidGenerator
        self._fragmentHandler: FragmentHandlerDefault = FragmentHandlerDefault()
        self._fragment_size = Constants.MAX_SIZE_DATA

        self._ioThreads = 1
        self._codecThreads = os.cpu_count()
        self._exchangeThreads = os.cpu_count() * 4

        self._readBufferSize = 1024 * 4  # 4k
        self._writeBufferSize = 1024 * 4  # 4k

        self._sslContext = None
        self._exchangeExecutor: Optional[ThreadPoolExecutor] = None

        self._idleTimeout = 60_000  # 60秒（心跳默认为20秒）
        self._requestTimeout = 10_000  # 10秒（默认与连接超时同）
        self._streamTimeout = 1000 * 60 * 60 * 2  # 2小时 //避免永不回调时，不能释放

        self._maxUdpSize = 2048

        self.__isThread = False
        self.__loggerLevel = "INFO"

    def __del__(self):
        if self._exchangeExecutor:
            self._exchangeExecutor.shutdown()

    def client_mode(self):
        return self._clientMode

    def is_serial_send(self) -> bool:
        return self._serialSend

    def serial_send(self, serialSend: bool):
        self._serialSend = serialSend
        return self

    def is_nolock_send(self) -> bool:
        return self._nolockSend

    def nolock_send(self, nolockSend: bool):
        self._nolockSend = nolockSend
        return self

    def get_stream_manger(self) -> StreamManger:
        return self._streamManger

    def get_role_name(self) -> str:
        return "Client" if self.client_mode() else "Server"

    def get_charset(self):
        return self._charset

    def charset(self, charset):
        self._charset = charset
        return self

    def get_codec(self):
        return self._codec

    def get_fragment_handler(self):
        return self._fragmentHandler

    def fragment_handler(self, fragmentHandler):
        Asserts.assert_null("fragmentHandler", fragmentHandler)

        self._fragmentHandler = fragmentHandler
        return self

    def get_fragment_size(self) -> int:
        return self._fragment_size

    def fragment_size(self, fragmentSize: int):
        if fragmentSize > Constants.MAX_SIZE_DATA:
            raise SocketDException("The parameter fragmentSize cannot > 16m")

        if fragmentSize < Constants.MIN_FRAGMENT_SIZE:
            raise SocketDException("The parameter fragmentSize cannot < 1k")

        self._fragment_size = fragmentSize
        return self

    def gen_id(self) -> str:
        return self._idGenerator()

    def id_generator(self, idGenerator: IdGenerator):
        Asserts.assert_null("idGenerator", idGenerator)

        self._idGenerator = idGenerator
        return self

    def get_ssl_context(self) -> ssl.SSLContext:
        return self._sslContext

    def ssl_context(self, sslContext: ssl.SSLContext):
        self._sslContext = sslContext
        return self

    def get_exchange_executor(self):
        if self._exchangeExecutor is None:
            __nThreads = self._exchangeThreads
            self._exchangeExecutor = ThreadPoolExecutor(max_workers=__nThreads,
                                                        thread_name_prefix="socketd-exchangeExecutor-")
        return self._exchangeExecutor

    def exchange_executor(self, exchangeExecutor):
        self._exchangeExecutor = exchangeExecutor
        return self

    def get_io_threads(self):
        return self._ioThreads

    def io_threads(self, ioThreads):
        self._ioThreads = ioThreads
        return self

    def get_codec_threads(self):
        return self._codecThreads

    def codec_threads(self, codecThreads):
        self._codecThreads = codecThreads
        return self

    def get_exchange_threads(self):
        return self._exchangeThreads

    def exchange_threads(self, exchangeThreads):
        self._exchangeThreads = exchangeThreads
        return self

    def get_read_buffer_size(self) -> int:
        return self._readBufferSize

    def read_buffer_size(self, read_buffer_size):
        self._readBufferSize = read_buffer_size
        return self

    def get_write_buffer_size(self) -> int:
        return self._writeBufferSize

    def write_buffer_size(self, write_buffer_size):
        self._writeBufferSize = write_buffer_size
        return self

    def get_idle_timeout(self) -> float:
        return self._idleTimeout

    def idle_timeout(self, _idle_timeout: float):
        self._idleTimeout = _idle_timeout
        return self

    def get_request_timeout(self) -> float:
        return self._requestTimeout

    def request_timeout(self, _request_time_out):
        self._requestTimeout = _request_time_out
        return self

    def get_stream_timeout(self) -> float:
        return self._streamTimeout

    def stream_timeout(self, _stream_timeout):
        self._streamTimeout = _stream_timeout
        return self

    def is_thread(self, _is_thread):
        self.__isThread = _is_thread
        return self

    def get_is_thread(self):
        return self.__isThread

    def get_logger_level(self) -> str:
        return self.__loggerLevel

    def logger_level(self, __logger_level: str):
        self.__loggerLevel = __logger_level
        logger.setLevel(__logger_level)
        return self

    def get_max_udp_size(self) -> int:
        return self._maxUdpSize

    def max_udp_size(self, maxUdpSize):
        self._maxUdpSize = maxUdpSize
        return self

File: javascript/socketd/src/socketd/transport/core/impl/ConfigBase.ts

import type {Codec} from "../Codec";
import {IdGenerator} from "../IdGenerator";
import {FragmentHandler} from "../FragmentHandler";
import {CodecDefault} from "../codec/CodecDefault";
import {Constants} from "../Constants";
import {Asserts} from "../Asserts";
import {Config} from "../Config";
import {FragmentHandlerDefault} from "../fragment/FragmentHandlerDefault";
import {GuidGenerator} from "../identifier/GuidGenerator";
import {StreamManger} from "../../stream/StreamManger";
import {StreamMangerDefault} from "../../stream/impl/StreamMangerDefault";


export abstract class ConfigBase implements Config {
    //是否客户端模式
    private _clientMode: boolean;
    //流管理器
    private _streamManger: StreamManger;
    //编解码器
    private _codec: Codec;

    //id生成器
    private _idGenerator: IdGenerator;
    //分片处理
    private _fragmentHandler: FragmentHandler;
    //分片大小
    private _fragmentSize: number;
    //字符集
    protected _charset: string
    //io线程数
    protected _ioThreads: number;
    //解码线程数
    protected _codecThreads: number;
    //交换线程数
    protected _exchangeThreads: number;
    //读缓冲大小
    protected _readBufferSize: number;
    //写缓冲大小
    protected _writeBufferSize: number;

    //连接空闲超时
    protected _idleTimeout: number;
    //请求默认超时
    protected _requestTimeout: number;
    //消息流超时（从发起到应答结束）
    protected _streamTimeout: number;
    //最大udp包大小
    protected _maxUdpSize: number;
    //使用子协议
    protected _useSubprotocols: boolean;

    constructor(clientMode: boolean) {
        this._clientMode = clientMode;
        this._streamManger = new StreamMangerDefault(this);
        this._codec = new CodecDefault(this);

        this._charset = "utf-8";

        this._idGenerator = new GuidGenerator();
        this._fragmentHandler = new FragmentHandlerDefault();
        this._fragmentSize = Constants.MAX_SIZE_DATA;

        this._ioThreads = 1;
        this._codecThreads = 2;
        this._exchangeThreads = this._codecThreads * 4;

        this._readBufferSize = 512;
        this._writeBufferSize = 512;

        this._idleTimeout = 60_000; //60秒
        this._requestTimeout = 10_000; //10秒（默认与连接超时同）
        this._streamTimeout = 1000 * 60 * 60 * 2;//2小时 //避免永不回调时，不能释放
        this._maxUdpSize = 2048; //2k //与 netty 保持一致 //实际可用 1464

        this._useSubprotocols = true;
    }

    /**
     * 是否客户端模式
     */
    clientMode(): boolean {
        return this._clientMode;
    }

    /**
     * 获取流管理器
     */
    getStreamManger(): StreamManger {
        return this._streamManger;
    }

    /**
     * 获取角色名
     * */
    getRoleName(): string {
        return this.clientMode() ? "Client" : "Server";
    }


    /**
     * 获取字符集
     */
    getCharset(): string {
        return this._charset;
    }

    /**
     * 配置字符集
     */
    charset(charset: string): this {
        this._charset = charset;
        return this;
    }

    /**
     * 获取编解码器
     */
    getCodec(): Codec {
        return this._codec;
    }

    /**
     * 获取标识生成器
     */
    genId(): string {
        return this._idGenerator.generate();
    }

    /**
     * 配置标识生成器
     */
    idGenerator(idGenerator: IdGenerator): this {
        Asserts.assertNull("idGenerator", idGenerator);

        this._idGenerator = idGenerator;
        return this;
    }

    /**
     * 获取分片处理
     */
    getFragmentHandler(): FragmentHandler {
        return this._fragmentHandler;
    }

    /**
     * 配置分片处理
     */
    fragmentHandler(fragmentHandler: FragmentHandler): this {
        Asserts.assertNull("fragmentHandler", fragmentHandler);

        this._fragmentHandler = fragmentHandler;
        return this;
    }

    /**
     * 获取分片大小
     */
    getFragmentSize(): number {
        return this._fragmentSize;
    }

    /**
     * 配置分片大小
     */
    fragmentSize(fragmentSize: number): this {
        if (fragmentSize > Constants.MAX_SIZE_DATA) {
            throw new Error("The parameter fragmentSize cannot > 16m");
        }

        if (fragmentSize < Constants.MIN_FRAGMENT_SIZE) {
            throw new Error("The parameter fragmentSize cannot < 1k");
        }

        this._fragmentSize = fragmentSize;
        return this;
    }

    /**
     * Io线程数
     * */
    getIoThreads(): number {
        return this._ioThreads;
    }

    /**
     * 配置Io线程数
     */
    ioThreads(ioThreads: number): this {
        this._ioThreads = ioThreads;
        return this;
    }

    /**
     * 获取解码线程数
     */
    getCodecThreads(): number {
        return this._codecThreads;
    }

    /**
     * 配置解码线程数
     */
    codecThreads(codecThreads: number): this {
        this._codecThreads = codecThreads;
        return this;
    }

    /**
     * 获取交换线程数
     */
    getExchangeThreads(): number {
        return this._exchangeThreads;
    }

    /**
     * 配置交换线程数
     */
    exchangeThreads(exchangeThreads: number): this {
        this._exchangeThreads = exchangeThreads;
        return this;
    }

    /**
     * 获取读缓冲大小
     */
    getReadBufferSize(): number {
        return this._readBufferSize;
    }

    /**
     * 配置读缓冲大小
     */
    readBufferSize(readBufferSize: number): this {
        this._readBufferSize = readBufferSize;
        return this;
    }

    /**
     * 获取写缓冲大小
     */
    getWriteBufferSize(): number {
        return this._writeBufferSize;
    }

    /**
     * 配置写缓冲大小
     */
    writeBufferSize(writeBufferSize: number): this {
        this._writeBufferSize = writeBufferSize;
        return this;
    }

    /**
     * 配置连接空闲超时
     */
    getIdleTimeout(): number {
        return this._idleTimeout;
    }

    /**
     * 配置连接空闲超时
     */
    idleTimeout(idleTimeout: number): this {
        this._idleTimeout = idleTimeout;
        return this;
    }

    /**
     * 配置请求默认超时
     */
    getRequestTimeout(): number {
        return this._requestTimeout;
    }

    /**
     * 配置请求默认超时
     */
    requestTimeout(requestTimeout: number): this {
        this._requestTimeout = requestTimeout;
        return this;
    }

    /**
     * 获取消息流超时（单位：毫秒）
     * */
    getStreamTimeout(): number {
        return this._streamTimeout;
    }

    /**
     * 配置消息流超时（单位：毫秒）
     * */
    streamTimeout(streamTimeout: number) {
        this._streamTimeout = streamTimeout;
        return this;
    }

    /**
     * 获取允许最大UDP包大小
     */
    getMaxUdpSize(): number {
        return this._maxUdpSize;
    }

    /**
     * 配置允许最大UDP包大小
     */
    maxUdpSize(maxUdpSize: number): this {
        this._maxUdpSize = maxUdpSize;
        return this;
    }

    /**
     * 生成 id
     * */
    generateId(): string {
        return this._idGenerator.generate();
    }

    /**
     * 配置使用子协议
     */
    useSubprotocols(useSubprotocols: boolean): this {
        this._useSubprotocols = useSubprotocols;
        return this;
    }

    /**
     * 是否使用子协议
     * */
    isUseSubprotocols(): boolean {
        return this._useSubprotocols;
    }
}